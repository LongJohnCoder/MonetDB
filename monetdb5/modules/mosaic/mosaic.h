/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB B.V.
 */

/* (c) M Kersten
 * The multi-purpose compressor infrastructure
 */
#ifndef _MOSLIST_H
#define _MOSLIST_H
#include "mal_interpreter.h"
#include "mal_exception.h"
#include "mal_function.h"

#include "mtime.h"
#include "math.h"
#include "opt_prelude.h"
#include "algebra.h"

//#define _DEBUG_MOSAIC_
//#define _DEBUG_PREFIX_

/* do not invest in compressing BATs smaller than this */
#define MOSAIC_THRESHOLD 1

/* The compressor kinds currently hardwired */
#define MOSAIC_METHODS	9
#define MOSAIC_RAW	0		// no compression at all
#define MOSAIC_RLE      1		// use run-length encoding
#define MOSAIC_CAPPED   2		// capped global dictionary encoding
#define MOSAIC_VAR      3		// variable global dictionary encoding
#define MOSAIC_DELTA	4		// use delta encoding
#define MOSAIC_LINEAR 	5		// use an encoding for a linear sequence
#define MOSAIC_FRAME	6		// delta dictionary for frame of reference value
#define MOSAIC_PREFIX	7		// prefix/postfix bitwise compression
#define MOSAIC_EOL	8		// marker for the last block

//Compression should have a significant reduction to apply.
#define COMPRESS_THRESHOLD 50   //percent
#define MOSAICINDEX 8  //> 2 elements

/*
 * The header is reserved for meta information, e.g. oid indices.
 * The block header encodes the information needed for the chunk decompressor
 */
typedef Heap *mosaic;	// compressed data is stored on a heap.

#define METHOD_NOT_AVAILABLE -1

typedef struct MOSAICHEADER{
	int version;
	union{
		bte sumbte;
		bit sumbit;
		sht sumsht;
		int sumint;
		oid sumoid;
		lng sumlng;
#ifdef HAVE_HGE
		hge sumhge;
#endif
		flt sumflt;
		dbl sumdbl;
	} checksum, checksum2;	// for validity checks
	int top; // TODO: rename to e.g. nblocks because it is the number of blocks
	// both dictionary and framebased compression require a global dictionary of frequent values
	// Their size is purposely topped 
	bte mask, bits, framebits;	// global compression type properties
	int dictsize;		// used by capped compression, it is a small table
	union{
		bte valbte[256];
		sht valsht[256];
		int valint[256];
		lng vallng[256];
		oid valoid[256];
		flt valflt[256];
		dbl valdbl[256];
#ifdef HAVE_HGE
		hge valhge[256];
#endif
	}dict;
	lng dictfreq[256];// keep track on their use
	// collect compression statistics for the particular task
	// A value of METHOD_NOT_AVAILABLE in blks or elms indicates that the corresponding method wasn't considered as candidate.
	flt ratio;	//compresion ratio
	lng blks[MOSAIC_METHODS]; // number of blks per method.
	lng elms[MOSAIC_METHODS]; // number of compressed values in all blocks for this method.
	bte bits_var;
	ulng pos_var;
	ulng length_var;
	bte bits_capped;
	ulng pos_capped;
	ulng length_capped;
} * MosaicHdr;



/* limit the number of elements to consider in a block
 * It should always be smaller then: ~(0377<<MOSshift)
*/

#define MOSAICMAXCNT 100000
/* allow for experiementation using different block sizes */

#define CNT_BITS 24

typedef struct MOSAICBLK{
	unsigned int tag:(32 - CNT_BITS), cnt:CNT_BITS;
} MosaicBlkRec, *MosaicBlk;

#define MOSgetTag(Blk) (Blk->tag)
#define MOSsetTag(Blk,Tag)  (Blk)->tag = Tag
#define MOSsetCnt(Blk,I) (assert(I <= MOSAICMAXCNT), (Blk)->cnt = (unsigned int)(I))
#define MOSgetCnt(Blk) (BUN)((Blk)->cnt)
#define MOSincCnt(Blk,I) (assert((Blk)->cnt +I <= MOSAICMAXCNT), (Blk)->cnt+= (unsigned int)(I))

/* The start of the encoding withing a Mosaic block */
#define MOScodevector(Task) (((char*) (Task)->blk)+ MosaicBlkSize)

/* Memory word alignement is type and platform dependent.
 * We use an encoding that fits the column type requirements
 */
#define wordaligned(SZ,TYPE) \
	 ((SZ) +  ((SZ) % sizeof(TYPE)? sizeof(TYPE) - ((SZ)%sizeof(TYPE)) : 0))

// alignment is focused on mosaichdr size
#define MosaicHdrSize  wordaligned(sizeof(struct MOSAICHEADER),sizeof(struct MOSAICBLK))
#define MosaicBlkSize  sizeof(MosaicBlkRec)

// helper function to check if a value is equal to the nil value of its type.
// TODO: check if this function can be added to gdk_atoms.h
#define is_nil(TPE, v) (ATOMcmp( TYPE_##TPE , (void*) &(v), (void*) ATOMnilptr(TYPE_##TPE) ) == 0)

#define getSrc(TPE, TASK) (((TPE*)TASK->src) + TASK->start)

typedef struct _GlobalCappedInfo GlobalCappedInfo;
typedef struct _GlobalVarInfo GlobalVarInfo;

/* The (de) compression task descriptor */
typedef struct MOSTASK{
	int type;		// one of the permissible compression types
	int filter[MOSAIC_METHODS];// algorithmic (de)compression mix

	/* collect the range being applied for the last compression call */
	BUN range[MOSAIC_METHODS]; // end of compression range
	float factor[MOSAIC_METHODS];// compression factor of last range. // TODO: This doesn't seem to be used. It also has some design flaws to it. Like what is its value if there are two blocks of the same compression type.

	MosaicHdr hdr;	// header block with index/synopsis information
	MosaicBlk blk;	// current block header in scan
	GlobalVarInfo* var_info;
	GlobalCappedInfo* capped_info;
	char 	 *dst;		// write pointer into current compressed blocks
	oid 	start;		// oid of first element in current blk
	oid		stop;		// last oid of range to be scanned

	BAT *bsrc;		// target column to extended with compressed heap
	char *src;		// read pointer into source

	lng timer;		// compression time

	oid *lb, *rb;	// Collected oids from operations
	oid *cl;		// candidate admin
	lng	n;			// element count in candidate list
	BUN cnt;		// elements in result set

	BAT *lbat, *rbat; // for the joins, where we dont know their size upfront

} *MOStask;

typedef struct _MosaicEstimation {
	BUN compressed_size;
	BUN uncompressed_size;
	MosaicBlkRec compression_strategy;
	bool is_applicable;
	bool must_be_merged_with_previous;
	size_t nr_var_encoded_elements;
	size_t nr_var_encoded_blocks;
	size_t nr_capped_encoded_elements;
	size_t nr_capped_encoded_blocks;
} MosaicEstimation;

/* Run through a column to produce a compressed version */

// skip until you hit a candidate
#define MOSskipit()\
if ( task->n && task->cl ){\
	while(task->n > 0 && *task->cl < (oid) first)\
		{task->cl++; task->n--;}\
	if (task->n <= 0 || *task->cl > (oid) first )\
		continue;\
	if ( *task->cl == (oid) first ){\
		task->cl++; task->n--; \
	}\
} else if (task->cl) continue;



mal_export char *MOSfiltername[];
mal_export bool MOSisTypeAllowed(int compression, BAT* b);
mal_export str MOScompress(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str MOSdecompress(bat* ret, const bat* bid);
mal_export str MOScompressInternal(BAT* bsrc, const char* compressions);
mal_export str MOSselect(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str MOSthetaselect(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str MOSprojection(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str MOSjoin(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export str MOSlayout(BAT *b, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export str MOSsliceInternal(bat *slices, BUN size, BAT *b);
mal_export void MOSAnalysis(BAT *b, BAT *btech, BAT *output, BAT *factor, BAT *compress, BAT *decompress, str compressions);
mal_export str MOSslice(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci);
mal_export void MOSblk(MosaicBlk blk);
mal_export BUN MOSlimit(void);

#endif /* _MOSLIST_H */
