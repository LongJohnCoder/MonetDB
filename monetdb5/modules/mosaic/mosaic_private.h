/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2019 MonetDB B.V.
 */

/* authors Martin Kersten, Aris Koning
 */

#define PREVVALUEbit(x) ((x) - 1)
#define PREVVALUEbte(x) ((x) - 1)
#define PREVVALUEsht(x) ((x) - 1)
#define PREVVALUEint(x) ((x) - 1)
#define PREVVALUElng(x) ((x) - 1)
#define PREVVALUEoid(x) ((x) - 1)
#ifdef HAVE_HGE
#define PREVVALUEhge(x) ((x) - 1)
#endif
#define PREVVALUEflt(x) nextafterf((x), -GDK_flt_max)
#define PREVVALUEdbl(x) nextafter((x), -GDK_dbl_max)
#define PREVVALUEdate(x) ((x) - 1)
#define PREVVALUEdaytime(x) ((x) - 1)

#define NEXTVALUEbit(x) ((x) + 1)
#define NEXTVALUEbte(x) ((x) + 1)
#define NEXTVALUEsht(x) ((x) + 1)
#define NEXTVALUEint(x) ((x) + 1)
#define NEXTVALUElng(x) ((x) + 1)
#define NEXTVALUEoid(x) ((x) + 1)
#ifdef HAVE_HGE
#define NEXTVALUEhge(x) ((x) + 1)
#endif
#define NEXTVALUEflt(x) nextafterf((x), GDK_flt_max)
#define NEXTVALUEdbl(x) nextafter((x), GDK_dbl_max)
#define NEXTVALUEdate(x) ((x) + 1)
#define NEXTVALUEdaytime(x) ((x) + 1)
