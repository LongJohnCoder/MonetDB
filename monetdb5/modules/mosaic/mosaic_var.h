/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB B.V.
 */

/*
 * (c)2014 author Martin Kersten
 */

#ifndef _MOSAIC_VAR_
#define _MOSAIC_VAR_

#include <mal.h>
#include "mal_interpreter.h"
#include "mal_client.h"
#include "mosaic_utility.h"

#define DICTTHRESHOLD 4192
#define DICTSIZE 256

bool MOStypes_var(BAT* b);
mal_export void MOScreatevar(MOStask task);
mal_export void MOSlayout_var(MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSlayout_var_hdr(MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties);
mal_export void MOSadvance_var(MOStask task);
mal_export void MOSskip_var(MOStask task);
mal_export str MOSprepareEstimate_var(MOStask task);
mal_export str  MOSestimate_var(MOStask task, MosaicEstimation* current, const MosaicEstimation* previous);
mal_export void MOSpostEstimate_var(MOStask task);
mal_export str finalizeDictionary_var(MOStask task);
mal_export void MOScompress_var(MOStask task, MosaicBlkRec* estimate);
mal_export void MOSdecompress_var(MOStask task);

ALGEBRA_INTERFACES_ALL_TYPES(var);

#define DO_OPERATION_ON_var(OPERATION, TPE) DO_OPERATION_ON_ALL_TYPES(OPERATION, var, TPE)

#endif /* _MOSAIC_VAR_ */
