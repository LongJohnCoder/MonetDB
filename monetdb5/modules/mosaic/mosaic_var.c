/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2018 MonetDB b.V.
 */


/*
 * 2014-2016 author Martin Kersten
 * Global info encoding
 * Index value zero is not used to easy detection of filler values
 * The info index size is derived from the number of entries covered.
 * It leads to a compact n-bit representation.
 * Floating points are not expected to be replicated 
 * A limit of 256 elements is currently assumed.
 */

#include "monetdb_config.h"
#include "gdk.h"
#include "gdk_bitvector.h"
#include "mosaic.h"
#include "mosaic_var.h"
#include "mosaic_private.h"
#include "mosaic_dictionary.h"

bool MOStypes_var(BAT* b) {
	switch (b->ttype){
	case TYPE_bit: return true; // Will be mapped to bte
	case TYPE_bte: return true;
	case TYPE_sht: return true;
	case TYPE_int: return true;
	case TYPE_lng: return true;
	case TYPE_oid: return true;
	case TYPE_flt: return true;
	case TYPE_dbl: return true;
#ifdef HAVE_HGE
	case TYPE_hge: return true;
#endif
	default:
		if (b->ttype == TYPE_date) {return true;} // Will be mapped to int
		if (b->ttype == TYPE_daytime) {return true;} // Will be mapped to lng
		if (b->ttype == TYPE_timestamp) {return true;} // Will be mapped to lng
	}
	return false;
}

void
MOSadvance_var(MOStask task)
{
	int *dst = (int*)  MOScodevectorDict(task);
	BUN cnt = MOSgetCnt(task->blk);
	long bytes;

	assert(cnt > 0);
	task->start += (oid) cnt;
	task->stop = task->stop;
	bytes =  (long) (cnt * task->hdr->bits_var)/8 + (((cnt * task->hdr->bits_var) %8) != 0);
	task->blk = (MosaicBlk) (((char*) dst)  + wordaligned(bytes, int)); 
}

typedef struct {
	MosaicBlkRec base;
	unsigned int bits;
	unsigned int dictsize;
	unsigned char offset; // padding for type alignment of the actual info entries.
} MOSDictHdr_t;

#define MOSgetDictFreq(DICTIONARY, KEY) ((ulng*)(((char*) DICTIONARY) + wordaligned(sizeof(DICTIONARY), ulng))[KEY])

typedef struct _GlobalVarInfo {
	BAT* dict;
	EstimationParameters parameters;
} GlobalVarInfo;

#define GET_BASE(INFO, TPE)				((TPE*) Tloc((INFO)->dict, 0))
#define GET_COUNT(INFO)					(BATcount((INFO)->dict))
#define GET_CAP(INFO)					(BATcapacity((INFO)->dict))
#define GET_DELTA_COUNT(INFO)				((INFO)->parameters.delta_count)
#define GET_BITS(INFO)					((INFO)->parameters.bits)
#define GET_BITS_EXTENDED(INFO)			((INFO)->parameters.bits_extended)
#define EXTEND(INFO, new_capacity)		(BATextend((INFO)->dict, new_capacity) == GDK_SUCCEED)
#define CONDITIONAL_INSERT(INFO, VAL, TPE)	(true)

// task dependent macro's
#define GET_FINAL_DICT(TASK, TPE) (((TPE*) (TASK)->bsrc->tvmosaic->base) + (TASK)->hdr->pos_var)
#define GET_FINAL_BITS(TASK) ((TASK)->hdr->bits_var)
#define GET_FINAL_DICT_COUNT(TASK) ((TASK)->hdr->length_var);\

#define DictionaryClass(TPE) \
find_value_DEF(TPE)\
insert_into_dict_DEF(TPE)\
extend_delta_DEF(TPE, GlobalVarInfo)\
merge_delta_Into_dictionary_DEF(TPE, GlobalVarInfo)\
compress_dictionary_DEF(TPE)\
decompress_dictionary_DEF(TPE)\
select_dictionary_DEF(TPE)\
thetaselect_dictionary_DEF(TPE)\
join_dictionary_DEF(TPE)

DictionaryClass(bte)
DictionaryClass(sht)
DictionaryClass(int)
DictionaryClass(oid)
DictionaryClass(lng)
DictionaryClass(flt)
DictionaryClass(dbl)
#ifdef HAVE_HGE
DictionaryClass(hge)
#endif

#define GetSizeInBytes(INFO) (BATcount((INFO)->dict) * (INFO)->dict->twidth)

void
MOSlayout_var_hdr(MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties)
{
	lng zero=0;
	unsigned int i;
	char buf[BUFSIZ];
	char bufv[BUFSIZ];
	(void) boutput;

	BUN dictsize = GET_COUNT(task->var_info);

	for(i=0; i< dictsize; i++){
		snprintf(buf, BUFSIZ,"var[%u]",i);
		if( BUNappend(btech, buf, false) != GDK_SUCCEED ||
			BUNappend(bcount, &zero, false) != GDK_SUCCEED ||
			BUNappend(binput, &zero, false) != GDK_SUCCEED ||
			// BUNappend(boutput, MOSgetDictFreq(dict_hdr, i), false) != GDK_SUCCEED ||
			BUNappend(bproperties, bufv, false) != GDK_SUCCEED)
		return;
	}
}

void
MOSlayout_var(MOStask task, BAT *btech, BAT *bcount, BAT *binput, BAT *boutput, BAT *bproperties)
{
	MosaicBlk blk = task->blk;
	lng cnt = MOSgetCnt(blk), input=0, output= 0;

	input = cnt * ATOMsize(task->type);
	output =  MosaicBlkSize + (cnt * task->hdr->bits_var)/8 + (((cnt * task->hdr->bits_var) %8) != 0);
	if( BUNappend(btech, "var blk", false) != GDK_SUCCEED ||
		BUNappend(bcount, &cnt, false) != GDK_SUCCEED ||
		BUNappend(binput, &input, false) != GDK_SUCCEED ||
		BUNappend(boutput, &output, false) != GDK_SUCCEED ||
		BUNappend(bproperties, "", false) != GDK_SUCCEED)
		return;
}

void
MOSskip_var(MOStask task)
{
	MOSadvance_var(task);
	if ( MOSgetTag(task->blk) == MOSAIC_EOL)
		task->blk = 0; // ENDOFLIST
}

str
MOSprepareEstimate_var(MOStask task)
{
	str error;

	GlobalVarInfo** info = &task->var_info;
	BAT* source = task->bsrc;

	if ( (*info = GDKmalloc(sizeof(GlobalVarInfo))) == NULL ) {
		throw(MAL,"mosaic.var",MAL_MALLOC_FAIL);	
	}

	BAT* dict;
	if ((dict = COLnew(0, source->ttype, 0, TRANSIENT)) == NULL) {
		error = createException(MAL, "mosaic.var.COLnew", GDK_EXCEPTION);
		return error;
	}

	(*info)->dict = dict;

	return MAL_SUCCEED;
}

#define estimateDict(TASK, CURRENT, TPE)\
do {\
	/*TODO*/\
	GlobalVarInfo* info = TASK->var_info;\
	BUN limit = (TASK)->stop - (TASK)->start > MOSAICMAXCNT? MOSAICMAXCNT: (TASK)->stop - (TASK)->start;\
	TPE* val = getSrc(TPE, (TASK));\
	BUN delta_count;\
	BUN nr_compressed;\
\
	size_t old_keys_size	= ((CURRENT)->nr_var_encoded_elements * GET_BITS_EXTENDED(info)) / CHAR_BIT;\
	size_t old_dict_size	= GET_COUNT(info) * sizeof(TPE);\
	size_t old_headers_size	= (CURRENT)->nr_var_encoded_blocks * (MosaicBlkSize + sizeof(TPE));\
	size_t old_bytes		= old_keys_size + old_dict_size + old_headers_size;\
\
	if (extend_delta_##TPE(&nr_compressed, &delta_count, limit, info, val)) {\
		throw(MAL, "mosaic.var", MAL_MALLOC_FAIL);\
	}\
\
	(CURRENT)->is_applicable = true;\
	(CURRENT)->nr_var_encoded_elements += nr_compressed;\
	(CURRENT)->nr_var_encoded_blocks++;\
\
	size_t new_keys_size	= ((CURRENT) -> nr_var_encoded_elements * GET_BITS_EXTENDED(info)) / CHAR_BIT;\
	size_t new_dict_size	= (delta_count + GET_COUNT(info)) * sizeof(TPE);\
	size_t new_headers_size	= (CURRENT)->nr_var_encoded_blocks * (MosaicBlkSize + sizeof(TPE));\
	size_t new_bytes		= new_keys_size + new_dict_size + new_headers_size;\
\
	(CURRENT)->compression_strategy.tag = MOSAIC_VAR;\
	(CURRENT)->compression_strategy.cnt = (unsigned int) nr_compressed;\
\
	(CURRENT)->uncompressed_size	+= (BUN) ( nr_compressed * sizeof(TPE));\
	(CURRENT)->compressed_size		+= (wordaligned( MosaicBlkSize, BitVector) + new_bytes - old_bytes);\
} while (0)

// calculate the expected reduction using INFO in terms of elements compressed
str
MOSestimate_var(MOStask task, MosaicEstimation* current, const MosaicEstimation* previous)
{	
	(void) previous;
	switch(ATOMbasetype(task->type)){
	case TYPE_bte: estimateDict(task, current, bte); break; 
	case TYPE_sht: estimateDict(task, current, sht); break;
	case TYPE_int: estimateDict(task, current, int); break;
	case TYPE_oid: estimateDict(task, current, oid); break;
	case TYPE_lng: estimateDict(task, current, lng); break;
	case TYPE_flt: estimateDict(task, current, flt); break;
	case TYPE_dbl: estimateDict(task, current, dbl); break;
#ifdef HAVE_HGE
	case TYPE_hge: estimateDict(task, current, hge); break;
#endif
	}
	return MAL_SUCCEED;
}

#define postEstimate(TASK, TPE) merge_delta_Into_dictionary_##TPE( (TASK)->var_info)

void
MOSpostEstimate_var(MOStask task) {
	switch(ATOMbasetype(task->type)){
	case TYPE_bte: postEstimate(task, bte); break; 
	case TYPE_sht: postEstimate(task, sht); break;
	case TYPE_int: postEstimate(task, int); break;
	case TYPE_oid: postEstimate(task, oid); break;
	case TYPE_lng: postEstimate(task, lng); break;
	case TYPE_flt: postEstimate(task, flt); break;
	case TYPE_dbl: postEstimate(task, dbl); break;
#ifdef HAVE_HGE
	case TYPE_hge: postEstimate(task, hge); break;
#endif
	}
}

static str
_finalizeDictionary(BAT* b, GlobalVarInfo* info, ulng* pos_dict, ulng* length_dict, bte* bits_dict) {
	Heap* vmh = b->tvmosaic;
	BUN size_in_bytes = vmh->free + GetSizeInBytes(info);
	if (HEAPextend(vmh, size_in_bytes, true) != GDK_SUCCEED) {
		throw(MAL, "mosaic.mergeDictionary_var.HEAPextend", GDK_EXCEPTION);
	}
	char* dst = vmh->base + vmh->free;
	char* src = info->dict->theap.base;
	/* TODO: consider optimizing this by swapping heaps instead of copying them.*/
	memcpy(dst, src, size_in_bytes);

	vmh->free += GetSizeInBytes(info);
	vmh->dirty = true;

	*pos_dict = 0;
	*length_dict = (ulng) GET_COUNT(info);
	*bits_dict = calculateBits((BUN) *length_dict);

	BBPreclaim(info->dict);

	GDKfree(info);

	return MAL_SUCCEED;
}

str
finalizeDictionary_var(MOStask task) {
	return _finalizeDictionary(
		task->bsrc,
		task->var_info,
		&task->hdr->pos_var,
		&task->hdr->length_var,
		&task->hdr->bits_var);
}

void
MOScompress_var(MOStask task, MosaicBlkRec* estimate)
{
	MosaicBlk blk = task->blk;
	task->dst = MOScodevectorDict(task);
	MOSsetTag(blk,MOSAIC_VAR);\

	switch(ATOMbasetype(task->type)){
	case TYPE_bte: DICTcompress(task, bte); break;
	case TYPE_sht: DICTcompress(task, sht); break;
	case TYPE_int: DICTcompress(task, int); break;
	case TYPE_lng: DICTcompress(task, lng); break;
	case TYPE_oid: DICTcompress(task, oid); break;
	case TYPE_flt: DICTcompress(task, flt); break;
	case TYPE_dbl: DICTcompress(task, dbl); break;
#ifdef HAVE_HGE
	case TYPE_hge: DICTcompress(task, hge); break;
#endif
	}
}

void
MOSdecompress_var(MOStask task)
{
	switch(ATOMbasetype(task->type)){
	case TYPE_bte: DICTdecompress(task, bte); break;
	case TYPE_sht: DICTdecompress(task, sht); break;
	case TYPE_int: DICTdecompress(task, int); break;
	case TYPE_lng: DICTdecompress(task, lng); break;
	case TYPE_oid: DICTdecompress(task, oid); break;
	case TYPE_flt: DICTdecompress(task, flt); break;
	case TYPE_dbl: DICTdecompress(task, dbl); break;
#ifdef HAVE_HGE
	case TYPE_hge: DICTdecompress(task, hge); break;
#endif
	}
}

str
MOSselect_var(MOStask task, void *low, void *hgh, bit *li, bit *hi, bit *anti)
{
	switch(ATOMbasetype(task->type)){
		case TYPE_bte: DICTselect(bte); break;
		case TYPE_sht: DICTselect(sht); break;
		case TYPE_int: DICTselect(int); break;
		case TYPE_lng: DICTselect(lng); break;
		case TYPE_oid: DICTselect(oid); break;
		case TYPE_flt: DICTselect(flt); break;
		case TYPE_dbl: DICTselect(dbl); break;
#ifdef HAVE_HGE
		case TYPE_hge: DICTselect(hge); break;
#endif
	}
	MOSskip_var(task);
	return MAL_SUCCEED;
}

str
MOSthetaselect_var( MOStask task, void *val, str oper)
{
	switch(ATOMbasetype(task->type)){
	case TYPE_bte: DICTthetaselect(bte); break;
	case TYPE_sht: DICTthetaselect(sht); break;
	case TYPE_int: DICTthetaselect(int); break;
	case TYPE_lng: DICTthetaselect(lng); break;
	case TYPE_oid: DICTthetaselect(oid); break;
	case TYPE_flt: DICTthetaselect(flt); break;
	case TYPE_dbl: DICTthetaselect(dbl); break;
#ifdef HAVE_HGE
	case TYPE_hge: DICTthetaselect(hge); break;
#endif
	}
	MOSskip_var(task);
	return MAL_SUCCEED;
}

str
MOSprojection_var( MOStask task)
{
	switch(ATOMbasetype(task->type)){
		case TYPE_bte: DICTprojection(bte); break;
		case TYPE_sht: DICTprojection(sht); break;
		case TYPE_int: DICTprojection(int); break;
		case TYPE_lng: DICTprojection(lng); break;
		case TYPE_oid: DICTprojection(oid); break;
		case TYPE_flt: DICTprojection(flt); break;
		case TYPE_dbl: DICTprojection(dbl); break;
#ifdef HAVE_HGE
		case TYPE_hge: DICTprojection(hge); break;
#endif
	}
	MOSskip_var(task);
	return MAL_SUCCEED;
}

str
MOSjoin_var( MOStask task, bit nil_matches)
{
	// set the oid range covered and advance scan range
	switch(ATOMbasetype(task->type)) {
		case TYPE_bte: DICTjoin(bte); break;
		case TYPE_sht: DICTjoin(sht); break;
		case TYPE_int: DICTjoin(int); break;
		case TYPE_lng: DICTjoin(lng); break;
		case TYPE_oid: DICTjoin(oid); break;
		case TYPE_flt: DICTjoin(flt); break;
		case TYPE_dbl: DICTjoin(dbl); break;
#ifdef HAVE_HGE
		case TYPE_hge: DICTjoin(hge); break;
#endif
	}
	MOSskip_var(task);
	return MAL_SUCCEED;
}
