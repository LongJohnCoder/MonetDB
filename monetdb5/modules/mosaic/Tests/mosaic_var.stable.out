stdout of test 'mosaic_var` in directory 'monetdb5/modules/mosaic` itself:


# 00:21:55 >  
# 00:21:55 >  "mserver5" "--debug=10" "--set" "gdk_nr_threads=0" "--set" "mapi_open=true" "--set" "mapi_port=34101" "--set" "mapi_usock=/var/tmp/mtest-14508/.s.monetdb.34101" "--set" "monet_prompt=" "--forcemito" "--set" "mal_listing=2" "--dbpath=/export/scratch1/mk/mosaic//Linux/var/MonetDB/mTests_monetdb5_modules_mal" "mosaic_dict.mal"
# 00:21:55 >  

# MonetDB 5 server v11.20.0
# This is an unreleased version
# Serving database 'mTests_monetdb5_modules_mal', using 8 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs and 128bit integers dynamically linked
# Found 15.591 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2014 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://vienna.ins.cwi.nl:34101/
# Listening for UNIX domain connection requests on mapi:monetdb:///var/tmp/mtest-14508/.s.monetdb.34101
# MonetDB/GIS module loaded
# MonetDB/SQL module loaded
# MonetDB/R   module loaded

# 11:50:48 >  
# 11:50:48 >  "mclient" "-lmal" "-ftest" "-tnone" "-Eutf-8" "--host=/var/tmp/mtest-30332" "--port=39586"
# 11:50:48 >  

#--------------------------#
# h	t  # name
# void	timestamp  # type
#--------------------------#
[ 0@0,	2014-08-14 12:00:00.000000	]
[ 1@0,	2014-08-14 12:00:00.000000	]
[ 2@0,	2014-08-14 12:01:00.000000	]
[ 3@0,	2014-08-14 12:02:00.000000	]
[ 4@0,	2014-08-14 12:00:00.000000	]
[ 5@0,	2014-08-14 12:00:00.000000	]
[ 6@0,	2014-08-14 12:00:00.000000	]
[ 7@0,	2014-08-14 12:01:00.000000	]
[ 8@0,	2014-08-14 12:02:00.000000	]
[ 9@0,	2014-08-14 12:00:00.000000	]
#--------------------------#
# h	t  # name
# void	timestamp  # type
#--------------------------#
[ 0@0,	2014-08-14 12:00:00.000000	]
[ 1@0,	2014-08-14 12:00:00.000000	]
[ 2@0,	2014-08-14 12:01:00.000000	]
[ 3@0,	2014-08-14 12:02:00.000000	]
[ 4@0,	2014-08-14 12:00:00.000000	]
[ 5@0,	2014-08-14 12:00:00.000000	]
[ 6@0,	2014-08-14 12:00:00.000000	]
[ 7@0,	2014-08-14 12:01:00.000000	]
[ 8@0,	2014-08-14 12:02:00.000000	]
[ 9@0,	2014-08-14 12:00:00.000000	]
#--------------------------#
# h	t  # name
# void	date  # type
#--------------------------#
[ 0@0,	2014-08-14  ]
[ 1@0,	2014-08-14  ]
[ 2@0,	2014-08-14  ]
[ 3@0,	2014-08-14  ]
[ 4@0,	2014-08-15  ]
[ 5@0,	2014-08-14  ]
[ 6@0,	2014-08-14  ]
[ 7@0,	2014-08-14  ]
[ 8@0,	2014-08-14  ]
[ 9@0,	2014-08-15  ]
# clk 3	sizes 40	200	500%	      0.20x	none	1	10 10	
#--------------------------#
# h	t  # name
# void	date  # type
#--------------------------#
[ 0@0,	2014-08-14  ]
[ 1@0,	2014-08-14  ]
[ 2@0,	2014-08-14  ]
[ 3@0,	2014-08-14  ]
[ 4@0,	2014-08-15  ]
[ 5@0,	2014-08-14  ]
[ 6@0,	2014-08-14  ]
[ 7@0,	2014-08-14  ]
[ 8@0,	2014-08-14  ]
[ 9@0,	2014-08-15  ]

# 00:21:55 >  
# 00:21:55 >  "Done."
# 00:21:55 >  

